Add notes on cross cluster CRLs (#18986)
hashicorp/vault: website/content/docs/secrets/pki/considerations.mdx

`/config/auto-tidy` endpoint. Both of these should be enabled to ensure
compatibility with the wider PKIX ecosystem and performance of the cluster.

## Spectrum of Revocation Support

> Starting with Vault 1.13, the PKI engine has the ability to support a

Reviewer: ```suggestion
Starting with Vault 1.13, the PKI secrets engine has the ability to support a
```

spectrum of cluster sizes and certificate revocation quantities.

For users with few revocations or who want a unified view and have the
inter-cluster bandwidth to support it, we recommend turning on auto
rebuilding of CRLs, cross-cluster revocation queues, and cross-cluster CRLs.
This allows all consumers of the CRLs to have the most accurate picture of
revocations, regardless of which cluster they talk to.

> If the unified CRL becomes to big for the underlying storage mechanism or

Reviewer: Did you mean...

```suggestion
If the unified CRL becomes too big for the underlying storage mechanism or
```

for a single host to build, we recommend relying on OCSP instead of CRLs.
These have much smaller storage entries, and the CRL `disabled` flag is
independent of `unified_crls`, allowing unified OCSP to remain.

However, when cross-cluster traffic becomes too high (or if CRLs are still
necessary in addition to OCSP), we recommend sharding the CRL between
different clusters. This has been the default behavior of Vault, but with
the introduction of per-cluster, templated AIA information, the leaf
certificate's Authority Information Access (AIA) info will point directly
to the cluster which issued it, allowing the correct CRL for this cert to
be identified by the application. This more correctly mimics the behavior
of [Let's Encrypt's CRL sharding](https://letsencrypt.org/2022/09/07/new-life-for-crls.html).

> This sharding behavior can also be used for OCSP, if the cross-cluster

Reviewer: ```suggestion
This sharding behavior can also be used for OCSP when the cross-cluster
```

Author: I think this is more of an `if`, not a `when`. Not all customers will have enough PKI demand on a given mount that they'll necessarily hit cross-cluster bandwidth limitations.

traffic for revocation entries becomes too high.

For users who wish to manage revocation manually, using the audit logs to
track certificate issuance would allow an external system to identify which
certificates were issued. These can be manually tracked for revocation, and
a [custom CRL can be built](/vault/api-docs/secret/pki#combine-crls-from-the-same-issuer)
using externally tracked revocations. This would allow usage of roles set to
`no_store=true`, so Vault is strictly used as an issuing authority and isn't
storing any certificates, issued or revoked. For the highest of revocation
volumes, this could be the best option.

> Notably, this last approach can either be used for creation of externally

Reviewer: ```suggestion
Notably, this last approach can either be used for the creation of externally
```

stored unified or sharded CRLs. If a single external unified CRL becomes
unreasonably large, each cluster's certificates could have AIA info point

> to an externally stored and maintained, sharded CRL. However, at this time
> Vault has no mechanism to sign OCSP requests.

Reviewer: ```suggestion
to an externally stored and maintained, sharded CRL. However, 
Vault has no mechanism to sign OCSP requests at this time.
```

## Issuer Subjects and CRLs

As noted on several [GitHub issues](https://github.com/hashicorp/vault/issues/10176),
Go's x509 library has an opinionated parsing and structuring mechanism for
certificate's Subjects. Issuers created within Vault are fine, but when using

> externally created CA certificates, note that these may not be parsed

Reviewer: ```suggestion
externally created CA certificates, these may not be parsed
```

correctly throughout all parts of the PKI. In particular, CRLs embed a
(modified) copy of the issuer name. This can be avoided by using OCSP to
track revocation, but note that performance characteristics are different
between OCSP and CRLs.

~> Note: As of Go 1.19 and Vault 1.12, Go correctly formats the CRL's issuer
   name and this notice does not apply.

## Automate Leaf Certificate Renewal

> As much as possible, for managing certificates for services at scale, it is
> best to automate renewal of certificates. Vault agent [has support for

Reviewer: ```suggestion
To manage certificates for services at scale, it is best to automate the
certificate renewal as much as possible. Vault Agent [has support for
```
